internal/pkg/slashing/check.go

```diff
@@ -92,21 +87,28 @@ func (s *ConsensusFaultChecker) VerifyConsensusFault(ctx context.Context, h1, h2
 		}
 	}
 	// Parent-grinding fault: one block’s parent is a tipset that provably should have included some block but does not.
-	// The provable case is that two blocks are mined in consecutive epochs and the later one does not include the
-	// earlier one as a parent.
-	// B3 must prove that the higher block (B2) has grandparent equal to B1's parent.
-	if b1.Height+1 == b2.Height && !b2.Parents.Has(b1.Cid()) && b2.Parents.Has(b3.Cid()) && b3.Parents.Equals(b1.Parents) {
-		fault = &runtime.ConsensusFault{
-			Target: b1.Miner,
-			Epoch:  b2.Height,
-			Type:   runtime.ConsensusFaultParentGrinding,
+	// The provable case is that two blocks are mined and the later one does not include the
+	// earlier one as a parent even though it could have.
+	// B3 must prove that the higher block (B2) could have been included in B1's tipset.
+	if len(extra) > 0 {
+		innerErr = encoding.Decode(extra, &b3)
+		if innerErr != nil {
+			return nil, errors.Wrapf(innerErr, "failed to decode extra")
+		}
+		if b1.Height == b3.Height && b3.Parents.Equals(b1.Parents) && !b2.Parents.Has(b1.Cid()) && b2.Parents.Has(b3.Cid()) {
```

Comment:
it makes sense that something should change to account for null blocks, but `b1.Height == b3.Height` is a straight up double-fork mining fault, so doesn't seem right yet.

These changes seem to change the role of blocks, so I think I'll need a new diagram to understand it.

Reply:
So it isn't a double-fork mining fault because b1 and b3 need not be from the same miner (if they were then I agree it is a double-fork mining fault). The case in question is shown below: the case where there is a null block mined between the parent block and the faulty block (whiich omitted the parent block). 

That is, in the below B2 should include B1 as a parent but does not. This is a provable fault. In the prior check by looking at `B2.height == b1.height + 1`, we would miss this. As is, we are ok with null blocks on either side (ancestor to B1, B3 or ahead of B2). `B3.height == b1.height && b3.Parents == b1.Parents`` are the checks to prove that they should have been in the same tipset.

```
             ┌ ─ ─ ─ ─      ┌────────┐      ┌ ─ ─ ─ ─       ┌────────┐
                      │     │B3,     │               │      │B2,     │
             │              │MinerC  │      │               │MinerA  │
     ┌──────▶         │────▶│        │─────▶         │─────▶│        │
     │       │              │        │      │               │        │
┌────────┐    ─ ─ ─ ─ ┘     └────────┘       ─ ─ ─ ─ ┘      └────────┘
│B0,     │
│MinerAny│
│        │
│        │
└────────┘   ┌ ─ ─ ─ ─      ┌────────┐
     │                │     │B1,     │
     │       │              │MinerA  │
     └──────▶         │─────▶        │
             │              │        │
              ─ ─ ─ ─ ┘     └────────┘
```

Reply:
Very helpful, thank you.

What about if there's no null before B3, but there still is before B1, so B1.Height <= B3.Height, but still B2.Height > B1.Height? Given the parent linkages, maybe the only height check we need is B2 > B1?

Reply:
It's a very good question. Here's a try:

I think the check does have to include `B1.height == B3.height` in that this guarantees they could have been in the same tipset. 

In the case where `B1.height < B3.height` (so take B1 is a child of B0 and B3 has a null block as in the diagram), you could imagine that the miner A puts out B1, then hears about a new Tipset that includes B3 (and is heavier than the one off B1), and so logically chooses to miner atop that, yielding B2. Valid

Conversely, if  `B1.height > B3.height` (so the reverse, where B3 is a direct child of B0, B1 isn't), you could imagine that MinerA simply didn't hear about B3's tipset in time and so mined B1, before receiving B3's tipset and mining B2 atop it.
With a very strict definition of network synchrony (that assumes that max delay is less than a block time), that would be a provable fault, but I think it's a dangerous assumption (yields attacks where I can delay someone's block reception and then slash them).

So I don't think `B2 > B1` suffices for provable deviations. What do you think of the above reasoning?

Reply:
SGTM.